components/submit-expense/useExpenseForm.ts

```diff
@@ -0,0 +1,399 @@
+      options.schema = options.schema.extend({
+        expenseTypeOption: z
+          .nativeEnum(ExpenseTypeOption)
+          .refine(v => query.data.account.supportedExpenseTypes.includes(expenseTypeFromOption(v)), {
+            message: 'expense type not supported',
+          }),
+      });
+      options.supportedExpenseTypes = query.data.account.supportedExpenseTypes;
+    }
+
+    if (query.data?.loggedInAccount) {
+      options.payoutProfiles = getPayoutProfiles(query.data.loggedInAccount);
+    }
+
+    options.supportedCurrencies = Object.values(Currency);
+    if (query.data?.payee && values.payoutMethodId && values.expenseTypeOption) {
+      const supportedCurrencies = getSupportedCurrencies(query.data?.account, {
+        payee: query.data?.payee,
+        payoutMethod: query.data?.payee?.payoutMethods?.find(p => p.id === values.payoutMethodId),
+        type: expenseTypeFromOption(values.expenseTypeOption),
+        currency: account.currency,
+      });
+
+      options.schema = options.schema.extend({
+        expenseCurrency: z.string().refine(v => supportedCurrencies.includes(v), {
+          message: `Currency must be one of: ${supportedCurrencies.join(',')}`,
+        }),
+      });
+
+      options.supportedCurrencies = supportedCurrencies;
+    }
+
+    options.schema = options.schema.extend({
+      expenseItems: z.array(
+        z.object({
+          description: z.string().min(1),
+          url:
+            expenseTypeFromOption(values.expenseTypeOption) === ExpenseType.RECEIPT
+              ? z.string().url()
+              : z.string().url().optional(),
+          date: z.string(),
+          amount: z.object({
+            valueInCents: z.number(),
+            currency: z.string().refine(v => Object.values(Currency).includes(v as Currency), {
+              message: `Currency must be one of: ${Object.values(Currency).join(',')}`,
+            }),
+          }),
+        }),
+      ),
+    });
+
+    options.allowExpenseItemAttachment = options.allowExpenseItemCurrencyChange =
+      expenseTypeFromOption(values.expenseTypeOption) === ExpenseType.RECEIPT;
+
+    return options;
+  } catch (err) {
+    return options;
+  }
+}
+
+export function useExpenseForm(opts: {
+  initialValues: Parameters<typeof useFormik<ExpenseFormValues>>['0']['initialValues'];
+  onSubmit: Parameters<typeof useFormik<ExpenseFormValues>>['0']['onSubmit'];
+}): ExpenseForm {
+  const apolloClient = useApolloClient();
+  const { LoggedInUser } = useLoggedInUser();
+  const [formOptions, setFormOptions] = React.useState<ExpenseFormOptions>({ schema: minimumSchema });
+
+  const expenseForm: ExpenseFormik = useFormik<ExpenseFormValues>({
+    initialValues: opts.initialValues,
+    async validate(values) {
+      const result = formOptions.schema.safeParse(values);
+      if (result.success === false) {
+        const errs = {};
+
+        for (const issue of result.error.issues) {
+          set(errs, issue.path, issue.message);
```

Comment:
@Betree this WIP gets the zod validation error messages into the form errors, they are just not the friendly error messages ATM.
Maybe refining the zod schema itself could yield our custom i18n messages?

Reply:
Using `useFormikZod`, you should get proper i18n messages out of the box from `errors` with standard validations.

When using `refine`, you can do something like:

```ts
const getSchema = intl => {
  const schema = z.object({
    myValue: z.string().refine(isUppercase, value => ({ 
      message: intl.formatMessage({ defaultMessage: '{value} must be upper' }, { value })
    }))
  })
}
```